base_invader/controllers/main.py

```diff
@@ -108,30 +28,58 @@ def _get_component_context(self):
+        """
+        Return the component that implements the methods of the requested
+        service.
+        :param service_name:
+        :return: an instance of invader.service component
+        """
         with request.backend.work_on(
-                model_name=request.backend._name,
-                **self._get_component_context()
+            request.backend._name,
+            **self._get_component_context()
         ) as work:
             service = work.component(usage=service_name)
-            start = datetime.now()
-            method = request.httprequest.method
-            if method == 'GET':
-                res = service.get(params)
-            elif method == 'POST':
-                res = service.create(params)
-            elif method == 'PUT':
-                res = service.update(params)
-            elif method == 'DELETE':
-                res = service.delete(params)
-            res = self.make_json_response(res)
-            _logger.info('Invader Response in %s', datetime.now() - start)
-            return res
-
-    def make_json_response(self, data, headers=None, cookies=None):
-        data = JSONEncoder().encode(data)
-        if headers is None:
-            headers = {}
-        headers['Content-Type'] = 'application/json'
-        return request.make_response(data, headers=headers, cookies=cookies)
+            yield service
+
+    @route([
+        ROOT_PATH + '<string:_service_name>/',
+        ROOT_PATH + '<string:_service_name>/search',
+        ROOT_PATH + '<string:_service_name>/<int:_id>/',
+        ROOT_PATH + '<string:_service_name>/<int:_id>/get'
+    ], methods=['GET'], auth="api_key", csrf=False)
+    def get(self, _service_name, _id=None, **params):
+        with self.service_component(_service_name) as service:
+            method_name = 'get' if _id else 'search'
+            res = service.dispatch(method_name, _id, **params)
+            return self.make_response(res)
+
+    @route([
+        ROOT_PATH + '<string:_service_name>/',
+        ROOT_PATH + '<string:_service_name>/<string:method_name>',
+        ROOT_PATH + '<string:_service_name>/<int:_id>/',
+        ROOT_PATH + '<string:_service_name>/<int:_id>/<string:method_name>'
+    ], methods=['POST', 'PUT', 'DELETE'], auth="api_key", csrf=False)
+    def modify(self, _service_name, _id=None, method_name=None, **params):
+        method = request.httprequest.method
+        if not method_name:
+            if _id:
+                method_name = 'delete' if method == 'DELETE' else 'update'
+            else:
+                method_name = 'create'
+        if method_name.startswith('_'):
+            _logger.error('Invader service called with an unallowed method '
+                          'name: %s.\n Method can\'t start with "-"',
+                          method_name)
+            raise BadRequest()
+        with self.service_component(_service_name) as service:
+            res = service.dispatch(method_name, _id, **params)
```

Comment:
I see no risk in doing it this way since the dispatch method is responsible of this validation https://github.com/akretion/odoo-shopinvader/pull/75/files#diff-140c71b38c5cc504e483d6bfe1aa826bR133. On the other side I don't see an advantage to pass the params as kwrags in place of a simple dict. I'll change.